### ISSUE TITLE
Daemon fails to join process_events select statements

### ISSUE THREAD

**Issue by loqpa** [osql/osql#70]
# Bug report

I was trying to build a query that would return information about process and his parent in one SQL statement. Here is a simplified version of it:
`SELECT proc.pid, proc.path, proc.cmdline, proc.parent, prnt.path, prnt.cmdline FROM process_events AS proc LEFT JOIN (SELECT * FROM process_events) AS prnt ON proc.parent=prnt.pid;`

This worked well while i was debugging it in shell mode (`sudo osqueryd --S`). 
However when i switched to running osquery in daemon mode(`sudo osqueryd`) no parent info were available in query results.
### What operating system and version are you using?
```version = 18.04.2 LTS (Bionic Beaver)
   build = 
platform = ubuntu```
### What version of osquery are you using?
```version = 3.3.2```
### What steps did you take to reproduce the issue?
Config in /etc/osquery:
```
{
    "options": {
        "schedule_splay_percent": 20, 
        "events_max": 5000, 
        "audit_allow_sockets": "true", 
        "audit_persist": "true", 
        "disable_events": "false", 
        "events_expiry": 5000, 
        "audit_allow_config": "true", 
        "disable_audit": "false", 
        "host_identifier": "hostname", 
        "logger_plugin": "filesystem",
    }, 
    "schedule": {
        "process_events":
	   {
      "query": "SELECT proc.pid, proc.path, proc.cmdline, proc.parent, prnt.path, prnt.cmdline FROM process_events AS proc LEFT JOIN (SELECT * FROM process_events) AS prnt ON proc.parent=prnt.pid;",
      "interval": 10,
      "removed": false
	 }
    }
}
```
Start daemon with `sudo osqueryd`. Check `/var/log/osquery/osqueryd.results.log`
### What did you expect to see?
```
[
   {
      "name":"process_events",
      "hostIdentifier":"loqpa-VirtualBox",
      "calendarTime":"Wed Jun 19 16:36:12 2019 UTC",
      "unixTime":1560962172,
      "epoch":0,
      "counter":0,
      "proc.cmdline":"sudo echo Osquery slack",
      "proc.parent":"25159",
      "proc.parent_cmdline":"bash",
      "proc.parent_pid":"/bin/bash",
      "proc.path":"/usr/bin/sudo",
      "proc.pid":"26884"
   },
   {
       "name":"process_events",
      "hostIdentifier":"loqpa-VirtualBox",
      "calendarTime":"Wed Jun 19 16:36:12 2019 UTC",
      "unixTime":1560962172,
      "epoch":0,
      "counter":0,
      "proc.cmdline":"echo Osquery slack",
      "proc.parent":"26884",
      **"proc.parent_cmdline":"sudo echo Osquery slack",
      "proc.parent_pid":"/usr/bin/sudo",**
      "proc.path":"/bin/echo",
      "proc.pid":"26885"
   }
]
```
### What did you see instead?
```
[
   {
      "name":"process_events",
      "hostIdentifier":"loqpa-VirtualBox",
      "calendarTime":"Wed Jun 19 16:36:12 2019 UTC",
      "unixTime":1560962172,
      "epoch":0,
      "counter":0,
      "columns":{
         "proc.cmdline":"sudo echo Osquery slack",
         "proc.parent":"25159",
         "proc.parent_cmdline":"bash",
         "proc.parent_pid":"/bin/bash",
         "proc.path":"/usr/bin/sudo",
         "proc.pid":"26857"
      },
      "action":"added"
   },
   {
      "name":"process_events",
      "hostIdentifier":"loqpa-VirtualBox",
      "calendarTime":"Wed Jun 19 16:36:12 2019 UTC",
      "unixTime":1560962172,
      "epoch":0,
      "counter":0,
      "columns":{
         "proc.cmdline":"echo Osquery slack",
         "proc.parent":"26857",
         **"proc.parent_cmdline":"",
         "proc.parent_pid":"",**
         "proc.path":"/bin/echo",
         "proc.pid":"26858"
      },
      "action":"added"
   }
]
```

**Comment by zwass** [comment]
@loqpa Does this still happen if you set `--events_optimize=false`?

**Comment by loqpa** [comment]
@zwass This does solve the `parent` fields issue, however now freshly started daemon output in `/var/log/osquery/osqueryd.results.log` contains a lot of data from commands that i ran like an hour ago and. Is this behavior intended?

**Comment by zwass** [comment]
Yeah, ultimately this seems tricky... What the `--events_optimize` flag does is record the timestamp of the most recent event queried, and then looking at more recent events the next time the query runs. I'm not sure how to support both ensuring that all results are available for the join and not retrieving the full set of results on each run.

**Comment by uptycs-nishant** [comment]
@loqpa , please check if this helps -

https://github.com/osql/osql/blob/master/docs/wiki/development/pubsub-framework.md

Basically using range based query to events table.

**Comment by loqpa** [comment]
@uptycs-nishant considering link you recommended how *should* this work with flag `--events_expiry` set to `1`?

**Comment by loqpa** [comment]
Adding `WHERE proc.time > strftime('%s','now') - 15;` to the end of my original query seems to solve the *main* issue.
The result looks almost perfect. However due to `schedule_splay_percent` flag i can not choose strict borders for `where` clause (`interval` is set to `10` and i will either get duplicates or may miss something). Maybe it is possible to disable this flag for specific query?

**Comment by zwass** [comment]
Setting the time like this is unfortunately not a reliable solution. Because of the way the osquery scheduler works, you can't be guaranteed that a query scheduled for every 60 seconds will actually run on anything close to that interval. The interval is counted as 60 seconds of the osquery process actually running, not wall time. If the computer sleeps for a full day, you might end up seeing the real interval being 24 hours + 60 seconds. (Funny enough, I discussed this in my QueryCon presentation today)

We already have functionality that tracks the time the query was last run by the scheduler... This is what --events_optimize uses to set the appropriate range. I wonder if we could expose a SQL function that would allow you to reference that time directly in the query. It seems to me this would allow a full table scan to happen for the join, while only returning results from the child processes that match the time filter.

**Comment by loqpa** [comment]
In this case, how should we proceed with this issue? Should we open something like `feature request` and close this one?
